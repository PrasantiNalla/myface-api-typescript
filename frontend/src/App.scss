$background-color: #252A38;
$darker-background: #151820;
$foreground: #DCDCDC;

#root {
  max-width: 1280px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
}

body {
  background-color: $background-color;
  // display:flex;
}


nav.navbar {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  background-color: $darker-background;
}

div.nav-items {
  display: flex;
  justify-content: space-evenly;
  color: $foreground;
  list-style: none;
  text-decoration: none;
  padding: 15px;
}

div.nav-items a {
  list-style: none;
  color: $foreground;
  text-decoration: none;
  padding: 15px;
  font-weight: bold;

}

div.nav-items a:hover {
  background-color: $foreground;
  color: $darker-background
}

h1,
h3 {
  color: $foreground;
}