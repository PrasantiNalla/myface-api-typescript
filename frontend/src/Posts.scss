div.posts-wrapper {
    width: 100%;
    height: 100%;
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: repeat(3, 1fr);
    grid-gap: 50px;
}

li.posts-item img {
    width: 100%;
    height: 200px;
    margin-top: 10px;
    padding: 0;
    position:relative;
}

ul.posts {
    white-space: nowrap;
    list-style-type: none;
    text-overflow: wrap;
    background: #151820;
    list-style-type: none;
    width: 100%;
    height: 100%;
}

li.posts-item {
    list-style-type: none;
    max-width: 200px;
    background: #151820;
    position: relative;
    color: white;
    text-overflow: ellipsis;
    // max-height: 200px;
    // max-width: 200px;
    // box-sizing:border-box;
}