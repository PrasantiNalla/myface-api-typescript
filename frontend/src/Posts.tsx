import { useEffect, useState } from "react";
import {PostModel} from "../../src/models/api/postModel"
import {Page} from "../../src/models/api/page"

export function Posts(){
    const [myData, setMyData] = useState([])

    useEffect(()=>{
        fetch("http://localhost:3001/posts")
            .then(response => response.json())
            .then(data=>setMyData(data.results))
    },[]);

    return (
           <div>
                {myData.map((post: PostModel)=>{
                    return (
                            <li key={post.id}>
                                <img src={post.imageUrl}/>
                                <div>{post.createdAt.toLocaleString()}</div>
                                <div>{post.message}</div>
                            </li>
                            
                    )
                })}
            </div>
    )

}

