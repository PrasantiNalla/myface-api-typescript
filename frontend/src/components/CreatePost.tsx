import { FormEvent, useState } from "react";
import { createPostApi } from "../api/api";
import { useNavigate } from "react-router-dom";
import { Link } from "react-router-dom";
import "../styles/CreatePost.scss"

export function CreatePost() {

    const [message, setMessage] = useState<string>("");
    const [imageUrl, setImageUrl] = useState<string>("");
    const [alert, setAlert] = useState(false);

    const [error, setError] = useState<string>("");

    const navigate = useNavigate();

    function handleSubmit(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();
        
        createPostApi(message, imageUrl)
            .then(()=>navigate("/posts/success"))
            .catch(e=>{setError(e.message)})
    }

    return (
        <>
            <h2>Create a New Post</h2>

            <p className="error">{error.length > 0? error:""}</p>
            <form onSubmit={(event) => handleSubmit(event)}>
              
                <label>Message:
                    <input
                        type="text"
                        name="message"
                        id="message"
                        onChange={e => setMessage(e.target.value)}
                    />
                </label>

                {/* can we check type from backend? */}
                <label>Image URL:
                    <input type="url"
                        id="imageUrl"
                        name="imageUrl"
                        onChange={e => setImageUrl(e.target.value)}
                    />
                </label>
                <button type="submit" onClick={()=>setAlert(!alert)}>Create Post</button>
            </form>
        </>
    )
}