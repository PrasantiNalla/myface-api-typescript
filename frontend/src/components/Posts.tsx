import { useEffect, useState } from "react";
import { PostModel } from "../../../src/models/api/postModel"
import { Link } from "react-router-dom";
import { like, dislike } from "../api/api";
import { createInteraction } from "../../../src/repos/interactionRepo"
import React from "react";
import { FormEvent } from "react";
import { useNavigate } from "react-router-dom";

export function Posts() {
    const [myData, setMyData] = useState([])
    const [currentPage, setCurrentPage] = useState(1)
    const [status, setStatus] = useState(false)
    const [error, setError] = useState("")
    const [PostId, setPostId] = useState(0)
    const [alert, setAlert] = useState(false);
    const navigate = useNavigate();

    useEffect(() => {
        fetch(`http://localhost:3001/posts/?page=${currentPage}&pageSize=9`)
            .then(response => response.json())
            .then(data => setMyData(data.results))
    }, [currentPage, status]);

    function handleSubmit(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();

        like(PostId)
            .then(() => navigate("/posts"))
            .catch(e => { setError(e.message) })
        dislike(PostId)
            .then(() => navigate("/posts"))
            .catch(e => { setError(e.message) })
    }

    function errorMsg() {
        if (error.length > 0) {
            return error
        }
    }

    return (

        <main className="posts">
            <h2>All Posts</h2>
            <div className="posts-wrapper">
                {myData.map((post: PostModel) => {
                    return (
                        <div >
                            <ul className="posts" >
                                <li key={post.id} className="posts-item">

                                    <img src={post.imageUrl}
                                        alt="This is a post image"
                                        onError={({ currentTarget }) => {
                                            currentTarget.onerror = null; // prevents looping
                                            currentTarget.src = "https://cdn-multicoat-com.sfo2.digitaloceanspaces.com/wp-content/uploads/2018/08/02232112/placeholder.jpg";
                                        }} />
                                    <div>{post.createdAt.toLocaleString()}</div>
                                    <div>{post.message}</div>
                                    <form onSubmit={(event) => handleSubmit(event)}>
                                        <button className="btn like" onClick={() => { like(post.id); setStatus(!status); setPostId(post.id); setAlert(!alert)}}>{post.likedBy.length} Like </button>&nbsp;&nbsp;
                                        <button className="btn dislike" onClick={() => { dislike(post.id); setStatus(!status); setPostId(post.id); setAlert(!alert) }}>{post.dislikedBy.length} Dislike  </button>
                                    </form>
                                </li>
                            </ul>

                        </div>
                    )

                })}</div>

            {/* interesting bug: if not currentPage +1, the first time we click next, the url will -1 */}
            <Link to={`/posts/?page=${currentPage + 1}&pageSize=10`}
                onClick={() => (setCurrentPage(currentPage + 1))}>
                Next
            </Link>
            {/* add a condition to check for currentPage value before clicking previous */}
            &nbsp;&nbsp;&nbsp;&nbsp;
            <Link to={`/posts/?page=${currentPage - 1}&pageSize=10`}
                onClick={() => (currentPage > 1 ? setCurrentPage(currentPage - 1) : currentPage)}>
                Previous
            </Link>
        </main>
    )

}

