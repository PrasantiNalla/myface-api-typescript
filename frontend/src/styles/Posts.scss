div.posts-wrapper {
    width: 100%;
    height: 100%;
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-gap: 50px;
}

li.posts-item img {
    width: 100%;
    height: 100%;
    position: relative;
}

li.posts-item {
    list-style-type: none;
    background: #151820;
    position: relative;
    color: white;
    text-overflow: ellipsis;
}

// ul.posts {
//     white-space: nowrap;
//     list-style-type: none;
//     text-overflow: wrap;
//     background: #151820;
//     list-style-type: none;
// }