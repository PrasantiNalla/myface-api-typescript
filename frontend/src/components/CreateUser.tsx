import { FormEvent, useState } from "react";
import { createUserApi } from "../api/api";
import "../styles/CreatePost.scss"
import { useNavigate } from "react-router-dom";

export function CreateUser() {
    const [name, setName] = useState<string>("");
    const [userName, setUserName] = useState<string>("");
    const [email, setEmail] = useState<string>("");
    const [profileImage, setProfileImage] = useState<string>("");
    const [coverImage, setCoverImage] = useState<string>("");

    const navigate = useNavigate();

    function handleSubmit(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();
        navigate("/users/success");
        createUserApi(name, userName, email, profileImage, coverImage);
    }

  
    return (
        <>
            <h2>Create a new user</h2>
            <form onSubmit={(event) => handleSubmit(event)}>
                <label>Name
                    <input
                        type="text"
                        name="name"
                        id="name"
                        required
                        onChange={e => setName(e.target.value)}
                         />
                </label>
                <br /><br />
                <label>Username
                    <input
                        type="text"
                        name="username"
                        id="username"
                        required
                        onChange={e => setUserName(e.target.value)}
                     />

                </label>
                <br /><br />
                <label> Email
                    <input
                        type="email"
                        name="email"
                        id="email"
                        required
                        onChange={e => setEmail(e.target.value)}
                         />
                </label>
                <br /><br />
                <label> Profile Image
                    <input type="url"
                        name="profile-image"
                        id="profile-image"
                        required
                        onChange={e => setProfileImage(e.target.value)}
                         />
                </label>
                <br /><br />
                <label> Cover Image
                    <input type="url"
                        name="cover-image"
                        id="cover-image"
                        required 
                        onChange={e => setCoverImage(e.target.value)}
                        />
                </label>
                <br /><br />
                <button type="submit" >Create User</button>
            </form>
        </>
    )
}