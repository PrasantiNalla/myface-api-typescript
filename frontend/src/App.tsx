
import './App.scss'
// import './Posts.scss'
import './Users.scss'
import './UserDetail.scss'
import { Posts } from './Posts'
import { Users } from "./Users"
import { UserDetail } from './UserDetail'
import { CreateUser } from './CreateUser'
import { CreatePost } from './CreatePost'
import { NewUserMsg } from './NewUserMsg'
import { NewPostMsg } from './NewPostMsg'
import { BrowserRouter as Router, Routes, Route, Link } from "react-router-dom";

function App() {
  return (
    <>
      <Router>
        <Routes>
          <Route path="*" element={
            <div>
              <Link to="/posts">  Posts </Link>
              <Link to="/users">  Users </Link>
              <Link to="/users/create">  + New User </Link>
              <Link to="/posts/create">  + New Post </Link>
            </div>
          } />

          <Route path="/posts" element={<Posts />} />
          <Route path="/users" element={<Users />} />
          <Route path="/users/:userId" element={<UserDetail />} />
          <Route path="/users/create" element={<CreateUser />} />
          <Route path="/posts/create" element={<CreatePost />} />
          <Route path="/posts/success" element={<NewPostMsg />} />
          <Route path="/users/success" element={<NewUserMsg />} />

        </Routes>
      </Router>
    </>

  )
}

export default App
