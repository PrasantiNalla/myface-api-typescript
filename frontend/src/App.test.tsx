import App from './App';
import { Posts } from './components/Posts';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router, Routes, Route, BrowserRouter} from "react-router-dom";

// Fix the lib problem by these commands:
// npm install --save-dev react-test-renderer
// npm install --save-dev react-test-renderer --legacy-peer-deps
// npm i --save-dev @types/react-test-renderer
import * as renderer from 'react-test-renderer';

//unit test case
test('Check third image equal to what we are expecting', () => {
    render(<Router><Posts /></Router>);
    const image = screen.getAllByRole('img')[3];
    expect(image).toHaveAttribute('src', 'https://www.bing.com/th?id=OIP.W8apTKARTqVhmLzMCisNmAHaEs&w=185&h=185&c=8&rs=1&qlt=90&o=6&pid=3.1&rm=2')
});

//snapshot test
it('renders correctly2', async () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
});

//component test
