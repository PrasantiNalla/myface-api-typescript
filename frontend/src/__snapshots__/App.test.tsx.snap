// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly2 1`] = `
[
  <nav
    className="navbar"
  >
    <div
      className="nav-items"
    >
      <li>
        <a
          href="/"
          onClick={[Function]}
        >
           Home 
        </a>
      </li>
      <li>
        <a
          href="/posts"
          onClick={[Function]}
        >
            Posts 
        </a>
      </li>
      <li>
        <a
          href="/users"
          onClick={[Function]}
        >
            Users 
        </a>
      </li>
      <li>
        <a
          href="/users/create"
          onClick={[Function]}
        >
            + New User 
        </a>
      </li>
      <li>
        <a
          href="/posts/create"
          onClick={[Function]}
        >
            + New Post 
        </a>
      </li>
    </div>
  </nav>,
  <h1>
    Welcome to MyFace!
  </h1>,
  <h3>
    Get started by creating a new user!
  </h3>,
]
`;
